### PR TITLE
4779 payment method display

### LINKED ISSUE #4779: Wrong payment method
## Description

As an admin I can see on this order:

https://www.openfoodfrance.org/admin/orders/R463586725/payments 

Hub ID: 942

That the buyer experienced 3 failed stripe payments and ended up confirming her order by cash.

Yet she sees a confirmation page and email saying she has paid thanks to stripe (paiement par carte bancaire):

Even though the status is correctly at "unpaid" the confirmation of the payment method leads the buyer to think she does not have to pay :/

## Steps to Reproduce
I can't reproduce it yet...

## Animated Gif/Screenshot

image

## Severity
bug-s2: a non-critical feature is broken, no workaround

https://github.com/openfoodfoundation/openfoodnetwork/wiki/Bug-severity
-->

## Your Environment
I'm trying to get more info on what's happened.

### PR DESCRIPTION
**Pull request by mkllnk**
#### What? Why?

Closes #4779

If we had multiple failed payments and then a successful payment, the
order confirmation was displaying the payment method of the first failed
payment. That was confusing and is now changed to the last payment
method.

#### What should we test?

Sanity check:

- Check out normally.
- Confirm the "Paid via" display in the order confirmation page.
- Confirm the "Paid via" display in the order confirmation emails (customer and enterprise).

New behaviour:

- Check out with a failing payment. I think there are certain credit card numbers for Stripe that generate a failure.
- Then check out with a cash payment method.
- Confirm the "Paid via" display in the order confirmation page.
- Confirm the "Paid via" display in the order confirmation emails (customer and enterprise).

#### Release notes

Fixed: Order confirmations now display the last used payment method instead of the first.

Changelog Category: Fixed

### CODE REVIEW

File: app/helpers/orders_helper.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+module OrdersHelper
+  def order_paid_via(order)
+    # `sort_by` avoids additional database queries when payments are loaded
+    # already. There is usually only one payment and this shouldn't cause
+    # any overhead compared to `order(:updated_at)`.
+    #
+    # Using `last` without sort is not deterministic.
+    order.payments.sort_by(&:updated_at).last.andand.payment_method
```
Comment (luisramos0) [inline app/helpers/orders_helper.rb]: I think this helper method should be called "order_payment_method" and the implementation should call a service (app/services/order_payment) with this code (except the payment_method that can stay here.
I am sure this code will be used in many places in the app. I want to use it already in the subs work I am doing :-)
Maybe I''d call the method "last":
```suggestion
    def order_payment_method
         OrderPayment.new(order).last.payment_method
    end
```

Comment (mkllnk) [reply app/helpers/orders_helper.rb]: I'm happy for you to do the refactoring when you want to use it. You can then see more easily why it has been done that way.

Do you need to find the last payment or also the payment method of the last payment?

Didn't we agree on service names like `OrderPaymentFinder`?

Comment (luisramos0) [reply app/helpers/orders_helper.rb]: yesk, I can do the refactor afterwards.
in StripeSCA it''s mostly fetching the last pending payment.
yes, OrderPaymentFinder is more in accordance with our convention :+1:

File: app/services/order_payment_finder.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+module OrderPaymentFinder
+  def self.last_payment_method(order)
+    # `sort_by` avoids additional database queries when payments are loaded
+    # already. There is usually only one payment and this shouldn't cause
+    # any overhead compared to `order(:updated_at)`.
+    #
+    # Using `last` without sort is not deterministic.
+    order.payments.max_by(&:updated_at)&.payment_method
```
Comment (luisramos0) [inline app/services/order_payment_finder.rb]: 😍

**Comment by filipefurtad0** [comment]
Hey @mkllnk,

I have reproduced the bug:
- Checked out with a failing payment, with different stripe cards, from:
https://stripe.com/docs/testing
- Then check out with a cash payment method, and observed the reported bug.

Then I have staged this in Katuma:
-> filipefurtad0 manually deployed build #5 of pull-request-4790 to katuma-staging  on Wednesday, Feb 19 2020 at 09:02

Performed the sanity check:
- Check out normally.
- Confirm the "Paid via" display in the order confirmation page.
- Confirm the "Paid via" display in the order confirmation emails (customer and enterprise).

It all worked as expected.

Then I tested the scenario in which:
- Check out with a failing Stripe payment
- Check out with a cash payment method.

- Looked ath the "Paid via" display in the order confirmation page and the order confirmation emails (customer and enterprise).

In this case, I observed that the both the order confirmation page and the conformation email still show the information referring to the failed Stripe payment, with indication of "not payed", which is typical for cash payments (pic 1):
image

Only after manually capturing the order and payment (as admin or superadmin), one gets the correct information on the invoice (pic 2):

image

So, in summary: perhaps this PR should still be improved? As I understood the situation in pic.1 should't occur. The correct information should be also displayed on the invoice before manually capturing the cash payment.

**Comment by mkllnk** [comment]
Thank you @filipefurtad0! I will continue my work...

**Comment by mkllnk** [comment]
I was wondering about the `updated_at` column at some point but didn't follow up on that. It shows how important integration tests are. The `updated_at` column is updated for all payments during checkout. We get something like this:
```rb
[
    [0] #<Spree::Payment:0x000055d3c810e5b8> {
                          :id => 3,
                      :amount => 55.0,
                    :order_id => 6,
                  :created_at => Thu, 20 Feb 2020 15:48:52 AEDT +11:00,
                  :updated_at => Thu, 20 Feb 2020 15:48:55 AEDT +11:00,
                   :source_id => nil,
                 :source_type => nil,
           :payment_method_id => 1,
                       :state => "checkout",
               :response_code => nil,
                :avs_response => nil,
                  :identifier => "TLPK2JM6",
           :cvv_response_code => nil,
        :cvv_response_message => nil
    },
    [1] #<Spree::Payment:0x000055d3c7fbab58> {
                          :id => 2,
                      :amount => 55.0,
                    :order_id => 6,
                  :created_at => Thu, 20 Feb 2020 15:48:39 AEDT +11:00,
                  :updated_at => Thu, 20 Feb 2020 15:48:55 AEDT +11:00,
                   :source_id => 1,
                 :source_type => "Spree::CreditCard",
           :payment_method_id => 9,
                       :state => "invalid",
               :response_code => nil,
                :avs_response => nil,
                  :identifier => "P5CZ2ME2",
           :cvv_response_code => nil,
        :cvv_response_message => nil
    }
]
```

I don't think it's worth it writing a feature spec for this. I will just go for changing it to `created_at`.

**Comment by mkllnk** [comment]
The last commit is new.

**Comment by filipefurtad0** [comment]
Hey!
I see there is one review on hold on this PR. Is this OK to test or should I wait for the third review? Ping @mkllnk @kristinalim 
Thanks :-)

**Comment by luisramos0** [comment]
I think you can trust the workflow state Filipe, test ready :+1:

**Comment by filipefurtad0** [comment]
Hey @mkllnk,
I proceeded with testing :-)

First, I deployed the last build of the master branch:
- to make a quick sanity-check >> all good here.
- and to reproduce the bug >> strangely, I could't seem to reproduce it... So, after attempting a payment with an invalid card - 4000000000000002 - one gets the expected "Your card was declined" message. Paying with cash, as an alternative, yields:
image

..instead of the bug itself - showing the last used payment method (Stripe should appear).

After that, I staged this PR, and everything looks exaclty the same. So the issue appears to be solved.

However, I've noticed the such payments keep the "Failed" payment state in the backoffice/admin section, allthough the email-receipts to the customer and enterprise already show the correct information. So, in the following cases:

A - a regular cash payment
B - a cash payment, after an attempt using a invalid card

One gets:

image

The  payment state "failed" dissapears once the payment is manually captured, so I guess this is not a major issue. I'll add the feedback needed just to make sure you've seen this, and move it to ready to go, upon your thumbs-up.

Thank you!

**Comment by mkllnk** [comment]
Thank you, @filipefurtad0! It's entirely possible not to see the bug in master because it was random. Now it shouldn't be random anymore and always be right.

The payment list in the backoffice is a different issue not affected by this pull request. You can create an issue for it if you think it's needed. But as far as I know, nobody noticed this so far.

**Comment by filipefurtad0** [comment]
Thanks for your feedback @mkllnk!
I'll check whether this bug is pre-existing and if not I'll create one. In any case, I don't regard this as a serious issue, for it may be easily solved by manually capturing the order. It does cause some confusion, at first. Thanks again ;-)
